### PR TITLE
Add more banking stage benchmarks

### PR DESCRIPTION
**Pull request by ckamm**
These cover different batch sizes and account lock contention scenarios.

Some initial measurements for a different number of non-vote banking stage threads:

```
                     batch sz    1 threads   2 threads   3 threads
all same account       192         136ms       269ms       216ms
independent batches    192         136ms       109ms       106ms
two contentions        192           8ms         6ms         6ms
no same accounts       192           7ms         6ms         5ms
all same account        12          45ms       436ms (??)  142ms
independent batches     12          45ms        37ms        26ms
two contentions         12          14ms        11ms         8ms
no same accounts        12          11ms         9ms         6ms
```

After review feedback the PR was adjusted to improve banking-bench instead:
- Add write-lock-contention option, replacing same_payer
- write-lock-contention also has a same-batch-only value, where
  contention happens only inside batches, not between them
- Rename num-threads to batches-per-iteration, which is closer to what
  it is actually doing.
- Add num-banking-threads as a new option
- Rename packets-per-chunk to packets-per-batch, because this is closer
  to what's happening; and it was previously confusing that num-chunks
  had little to do with packets-per-chunk.

Example output for a iterations=100 and a permutation of inputs:

```
contention,threads,batchsize,batchcount,tps
none,           3,192, 4,65290.30
none,           4,192, 4,77358.06
none,           5,192, 4,86436.65
none,           3, 12,64,43944.57
none,           4, 12,64,65852.15
none,           5, 12,64,70674.37
same-batch-only,3,192, 4,3928.21
same-batch-only,4,192, 4,6460.15
same-batch-only,5,192, 4,7242.85
same-batch-only,3, 12,64,11377.58
same-batch-only,4, 12,64,19582.79
same-batch-only,5, 12,64,24648.45
full,           3,192, 4,3914.26
full,           4,192, 4,2102.99
full,           5,192, 4,3041.87
full,           3, 12,64,11316.17
full,           4, 12,64,2224.99
full,           5, 12,64,5240.32
```

#### Problem
Banking stage benchmarks cover only cases without account lock contention and one batch size.

#### Summary of Changes
Add benchmarks.

### CODE REVIEW

**Comment by ckamm** [comment]
@taozhu-chicago Is this something you're interested in reviewing?

The main issue I see is that the new benchmarks can take a good while to execute. test::Bencher seems to run each function at least 300 times, meaning something that takes 400ms per iteration takes 120s to benchmark.

**Comment by taozhu-chicago** [comment]
> @taozhu-chicago Is this something you're interested in reviewing?
> 
> The main issue I see is that the new benchmarks can take a good while to execute. test::Bencher seems to run each function at least 300 times, meaning something that takes 400ms per iteration takes 120s to benchmark.

Thanks for adding these benchmarks, happy to look at it, maybe later today. Just a thought (without looking into details), would these scenarios fit better in `banking_bench`?

**Comment by ckamm** [comment]
> Thanks for adding these benchmarks, happy to look at it, maybe later today. Just a thought (without looking into details), would these scenarios fit better in `banking_bench`?

Hah, I didn't know about banking_bench, will take a look tomorrow.

File: core/benches/banking_stage.rs [hunk 1]
```diff
@@ -167,7 +200,7 @@ fn bench_banking(bencher: &mut Bencher, tx_type: TransactionType) {
 
     // Set a high ticks_per_slot so we don't run out of ticks
     // during the benchmark
-    genesis_config.ticks_per_slot = 10_000;
+    genesis_config.ticks_per_slot = 100_000;
```
Comment (taozhu-chicago) [inline core/benches/banking_stage.rs:203]: not sure why want to increase this?

File: core/benches/banking_stage.rs [hunk 2]
```diff
@@ -287,14 +332,75 @@ fn bench_banking(bencher: &mut Bencher, tx_type: TransactionType) {
...
 }
 
 #[bench]
-fn bench_banking_stage_multi_programs(bencher: &mut Bencher) {
-    bench_banking(bencher, TransactionType::Programs);
+fn bench_banking_stage_batch_size_192_multi_accounts_contention_full(bencher: &mut Bencher) {
+    bench_banking_scaled(bencher, 192, TransactionType::AccountsContentionFull);
+}
+
+#[bench]
```
Comment (taozhu-chicago) [inline core/benches/banking_stage.rs:356]: Love to have these tests. May need to adjust CI if it's timed out due to additional benching

**Comment by taozhu-chicago** [comment]
> 

Actually, it looks better to add benches here as is

**Comment by ckamm** [comment]
> Actually, it looks better to add benches here as is

Are you sure? I was about to port things to banking-bench because that seemed more flexible and I wouldn't need to put all the permutations of options into the code. Do you disagree and would prefer to have it stay like this?

**Comment by sakridge** [comment]
> > Actually, it looks better to add benches here as is
> 
> Are you sure? I was about to port things to banking-bench because that seemed more flexible and I wouldn't need to put all the permutations of options into the code. Do you disagree and would prefer to have it stay like this?

I prefer banking-bench as well actually.

File: banking-bench/src/main.rs [hunk 3]
```diff
@@ -66,23 +66,55 @@ fn check_txs(
...
     (0..total_num_transactions)
         .into_par_iter()
-        .map(|_| {
+        .map(|i| {
             let mut new = dummy.clone();
             let sig: Vec<u8> = (0..64).map(|_| thread_rng().gen::<u8>()).collect();
-            if !same_payer {
-                new.message.account_keys[0] = solana_sdk::pubkey::new_rand();
-            }
-            new.message.account_keys[1] = solana_sdk::pubkey::new_rand();
```
Comment (taozhu-chicago) [inline banking-bench/src/main.rs:85]: +1 for replacing `same_payer` with contention on to account

**Review by taozhu-chicago — APPROVED** [review]
lgtm

**Pull request merged**